proxy/src/main/java/com/wavefront/agent/listeners/ZipkinPortUnificationHandler.java

```diff
@@ -0,0 +1,260 @@
+    if (uri == null) return;
+
+    String path = uri.getPath().endsWith("/") ? uri.getPath() : uri.getPath() + "/";
+
+    // Validate Uri Path and HTTP method of incoming Zipkin spans.
+    if (!ZIPKIN_VALID_PATHS.contains(path)) {
+      writeHttpResponse(ctx, HttpResponseStatus.BAD_REQUEST, "Unsupported URL path.", incomingRequest);
+      logWarning("WF-400: Requested URI path '" + path + "' is not supported.", null, ctx);
+      return;
+    }
+    if (!incomingRequest.method().toString().equalsIgnoreCase(ZIPKIN_VALID_HTTP_METHOD)) {
+      writeHttpResponse(ctx, HttpResponseStatus.BAD_REQUEST, "Unsupported Http method.", incomingRequest);
+      logWarning("WF-400: Requested http method '" + incomingRequest.method().toString() + "' is not supported.", null, ctx);
+      return;
+    }
+
+    HttpResponseStatus status;
+    StringBuilder output = new StringBuilder();
+    try {
+      byte[] bytesArray = new byte[incomingRequest.content().nioBuffer().remaining()];
+      incomingRequest.content().nioBuffer().get(bytesArray, 0, bytesArray.length);
+      BytesDecoder<zipkin2.Span> decoder = SpanBytesDecoderDetector.decoderForListMessage(bytesArray);
+      List<zipkin2.Span> zipkinSpanSink = new ArrayList<>();
+      decoder.decodeList(bytesArray, zipkinSpanSink);
+      processZipkinSpans(zipkinSpanSink);
+      status = HttpResponseStatus.ACCEPTED;
+      processedBatches.inc();
+    } catch (Exception e) {
+      failedBatches.inc();
+      writeExceptionText(e, output);
+      status = HttpResponseStatus.BAD_REQUEST;
+      logger.log(Level.WARNING, "Zipkin Thrift batch processing failed", Throwables.getRootCause(e));
+    }
+    writeHttpResponse(ctx, status, output, incomingRequest);
+  }
+
+  private void processZipkinSpans(List<zipkin2.Span> zipkinSpans) {
+    if (traceDisabled.get()) {
+      if (warningLoggerRateLimiter.tryAcquire()) {
+        logger.info("Ingested spans discarded because tracing feature is not enabled on the server");
+      }
+      discardedBatches.inc();
+      discardedTraces.inc(zipkinSpans.size());
+      return;
+    }
+
+    for (zipkin2.Span zipkinSpan : zipkinSpans) {
+      processZipkinSpan(zipkinSpan);
+    }
+  }
+
+  private void processZipkinSpan(zipkin2.Span zipkinSpan) {
+    List<Annotation> annotations = addAnnotations(zipkinSpan);
+
+    // TODO: Confirm source value
+    String sourceName = zipkinSpan.tags().get(SOURCE_KEY) == null ? "unknown" : zipkinSpan.tags().get(SOURCE_KEY);
+    annotations.add(new Annotation(SOURCE_KEY, sourceName));
+
+    //Build wavefront span
+    Span newSpan = Span.newBuilder()
+            .setCustomer("dummy")
+            .setName(zipkinSpan.name())
+            .setSource(sourceName)
+            .setSpanId(getSpanUuid(zipkinSpan))
+            .setTraceId(convertToUuidString(zipkinSpan.traceId()))
+            .setStartMillis(zipkinSpan.timestampAsLong() / 1000)
+            .setDuration(zipkinSpan.durationAsLong() / 1000)
```

Comment:
I thought you mentioned duration is in nanos?
if yes, then you need to convert it into millis not micros

Reply:
Mentioned incorrectly in the doc. Code does the right thing. (both timestamp and duration are reported by zipkin in micros)